File: src/test/java/org/luxons/sevenwonders/game/effects/ProductionIncreaseTest.java
package org.luxons.sevenwonders.game.effects;

import org.junit.experimental.theories.DataPoints;
import org.junit.experimental.theories.Theories;
import org.junit.experimental.theories.Theory;
import org.junit.runner.RunWith;
import org.luxons.sevenwonders.game.boards.Board;
import org.luxons.sevenwonders.game.resources.ResourceType;
import org.luxons.sevenwonders.game.resources.Resources;
import org.luxons.sevenwonders.game.test.TestUtils;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

@RunWith(Theories.class)
public class ProductionIncreaseTest {

    @DataPoints
    public static ResourceType[] resourceTypes() {
        return ResourceType.values();
    }

    private static ProductionIncrease createProductionIncrease(ResourceType... types) {
        ProductionIncrease effect = new ProductionIncrease();
        effect.getProduction().addAll(TestUtils.createFixedProduction(types));
        return effect;
    }

    @Theory
    public void apply_boardContainsAddedResourceType(ResourceType initialType, ResourceType addedType, ResourceType extraType) {
        Board board = TestUtils.createBoard(initialType);
        ProductionIncrease effect = createProductionIncrease(addedType);

        effect.apply(board, null, null);

        Resources resources = TestUtils.createResources(initialType, addedType);
        assertTrue(board.getProduction().contains(resources));

        Resources moreResources = TestUtils.createResources(initialType, addedType, extraType);
        assertFalse(board.getProduction().contains(moreResources));
    }
}

File: sw-engine/src/test/kotlin/org/luxons/sevenwonders/engine/effects/ProductionIncreaseTest.kt
package org.luxons.sevenwonders.engine.effects

import org.junit.experimental.theories.DataPoints
import org.junit.experimental.theories.Theories
import org.junit.experimental.theories.Theory
import org.junit.runner.RunWith
import org.luxons.sevenwonders.engine.SimplePlayer
import org.luxons.sevenwonders.model.resources.ResourceType
import org.luxons.sevenwonders.engine.resources.resourcesOf
import org.luxons.sevenwonders.engine.test.fixedProduction
import org.luxons.sevenwonders.engine.test.testBoard
import org.luxons.sevenwonders.engine.test.testTable
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

@RunWith(Theories::class)
class ProductionIncreaseTest {

    @Theory
    fun apply_boardContainsAddedResourceType(
        initialType: ResourceType,
        addedType: ResourceType,
        extraType: ResourceType
    ) {
        val board = testBoard(initialType)
        val effect = ProductionIncrease(fixedProduction(addedType), false)

        effect.applyTo(board)

        val resources = resourcesOf(initialType, addedType)
        assertTrue(board.production.contains(resources))
        assertFalse(board.publicProduction.contains(resources))

        val moreResources = resourcesOf(initialType, addedType, extraType)
        assertFalse(board.production.contains(moreResources))
        assertFalse(board.publicProduction.contains(moreResources))
    }

    @Theory
    fun apply_boardContainsAddedResourceType_sellable(
        initialType: ResourceType,
        addedType: ResourceType,
        extraType: ResourceType
    ) {
        val board = testBoard(initialType)
        val effect = ProductionIncrease(fixedProduction(addedType), true)

        effect.applyTo(board)

        val resources = resourcesOf(initialType, addedType)
        assertTrue(board.production.contains(resources))
        assertTrue(board.publicProduction.contains(resources))

        val moreResources = resourcesOf(initialType, addedType, extraType)
        assertFalse(board.production.contains(moreResources))
        assertFalse(board.publicProduction.contains(moreResources))
    }

    @Theory
    fun computePoints_isAlwaysZero(addedType: ResourceType) {
        val effect = ProductionIncrease(fixedProduction(addedType), false)
        val player = SimplePlayer(0, testTable(5))
        assertEquals(0, effect.computePoints(player))
    }

    companion object {

        @JvmStatic
        @DataPoints
        fun resourceTypes(): Array<ResourceType> = ResourceType.values()
    }
}
